src/trails.erl

```diff
@@ -151,18 +153,61 @@ trails(Handler) ->
   trails([Handler], []).
 
 %% @doc Store the given list of trails.
--spec store(trails()) -> ok.
+-spec store(Trails::trails() |
+              [{HostMatch::route_match(), Trails::trails()}]) -> ok.
 store(Trails) ->
   _ = application:ensure_all_started(trails),
+  store('_', Trails).
+
+-spec store(Server::ranch:ref(),
+            Trails::trails() |
+              [{HostMatch::route_match(), Trails::trails()}]) -> ok.
+store(_Server, []) ->
+  ok;
+store(Server, [{HostMatch, Trails} | Hosts]) ->
+  NormalizedPaths = normalize_store_input(Trails),
+  do_store(Server, HostMatch, NormalizedPaths),
+  store(Server, Hosts);
+store(Server, Trails) when is_list(Trails) ->
   NormalizedPaths = normalize_store_input(Trails),
-  store1(NormalizedPaths).
+  do_store(Server, '_', NormalizedPaths).
 
 %% @doc Retrieves all stored trails.
 -spec all() -> [trail()].
 all() ->
+    all('_').
+
+%% @doc Retrieves all stored trails for the given `HostMatch`
+-spec all(HostMatch::route_match()) -> [trail()].
+all(HostMatch) ->
+  all('_', HostMatch).
+
+%% @doc Retrieves all stored trails for the given `Server` and `HostMatch`
+-spec all(Server::ranch:ref(), HostMatch::route_match()) -> [trail()].
+all(Server, HostMatch) ->
   case application:get_application(trails) of
     {ok, trails} ->
-      Trails = all(ets:select(trails, [{{'$1', '$2'}, [], ['$$']}]), []),
+      MatchSpec =
+        case {Server, HostMatch} of
+          {'_', HostMatch} ->
+            [{{{'$1', HostMatch, '$2'}, '$3'}, [], ['$$']}];
+          {Server, HostMatch} ->
+            [{{{Server, HostMatch, '$1'}, '$2'}, [], ['$$']}]
+        end,
+      Matches = ets:select(trails, MatchSpec),
+      FoundServers = [Srvr ||
+                      Result = [Srvr, _PathMatch, _Trail] <-
+                      Matches, length(Result) > 2],
```

Comment:
since `Result` is matched against a list of 3 elements (`[Srvr, _PathMatch, _Trail]`), its length will always be greater than 2. There is no need for `length(Result) > 2` here.